GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs

```diff
@@ -545,53 +606,67 @@ static void StraightenLanes(int startIndex, int lastStraightenIndex, int lastLoo
-            // Then the cache is invalidated by setting _reorder.
+            // Then the cache is invalidated by setting _orderedNodesCache.
             // So if Sort() complains in this case, try again.
             while (true)
             {
                 try
                 {
-                    // Reset the reorder flag and the orderedUntilScore. This makes sure it isn't marked dirty before we even got to
-                    // rebuilding it.
-                    _orderedUntilScore = int.MinValue;
-                    _reorder = false;
-
-                    // Use a local variable, because the cached list can be reset.
-                    RevisionGraphRevision[] localOrderedNodesCache = _nodes.ToArray();
-                    Array.Sort(localOrderedNodesCache, (x, y) => x.Score.CompareTo(y.Score));
-                    _orderedNodesCache = localOrderedNodesCache;
-                    if (localOrderedNodesCache.Length > 0)
-                    {
-                        _orderedUntilScore = localOrderedNodesCache[^1].Score;
-                    }
+                    // Reset flags to ensure the cache isn't marked dirty before we even got to rebuilding it.
+                    _orderedNodesCacheInvalid = false;
+
+                    _orderedNodesCache = _revisionByObjectId.Values.OrderBy(n => n.Score).ToImmutableArray();
 
-                    return localOrderedNodesCache;
+                    return _orderedNodesCache;
                 }
-                catch (ArgumentException ex) when (_reorder && ex.Message.Contains("IComparer.Compare()"))
+                catch (ArgumentException ex) when (_orderedNodesCacheInvalid && ex.Message.Contains("IComparer.Compare()"))
                 {
                     // ignore and try again
                 }
             }
         }
 
         /// <summary>
-        /// Mark the cache as invalid if the node scores have been changed.
+        /// Mark the node and row cache as invalid if the node scores in the ordered node cache have been changed.
         /// </summary>
         /// <param name="minScore">The minimum score of the modified nodes.</param>
-        private void MarkCacheAsInvalidIfNeeded(int minScore)
+        /// <returns><see langword="true"/> if the row cache is invalidated; otherwise <see langword="false"/>.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool MarkCacheAsInvalidIfNeeded(int minScore)
         {
-            if (minScore <= _orderedUntilScore)
+            if (!_orderedNodesCacheInvalid)
             {
-                _reorder = true;
+                ImmutableArray<RevisionGraphRevision> localOrderedNodesCache = _orderedNodesCache;
+                if (localOrderedNodesCache.Length > 0 && minScore <= localOrderedNodesCache[^1].Score)
+                {
+                    _orderedNodesCacheInvalid = true;
+                }
             }
```

Comment:
I have analyzed this moved code multiple times. I do still not see why it causes an empty revision grid for many seconds when switching to the Linux repo.